### DISCUSSION TITLE
How to send data to a website, what would be a general approach?

### DISCUSSION THREAD

**Discussion by PetersOtto** [datenstrom/community#868] [Ask a question] [answered]
Hello

I have a question that is not only related to Yellow. I hope it's ok if I ask it here.

I have a Yellow website outside my home network.
I also use a Raspberry Pi Pico W to collect sensor data (inside my home network). This data are currently sent to me by the Pico via email.

I want the Pico to send these data directly to my Yellow website.

I have no idea how to do this. Can you think of anything?

**Comment by schulle4u** [comment]
I'm afraid there's no general solution without further information about the data, the format etc. If this data is stored in e.g csv files you could write a shell script which automatically uploads the file via FTP/SFTP from the Pi to your website, then use an appropriate file parser to display them as needed. You could use My CSV extension, or have a look at the extensions by @GiovanniSalmeri, he has some parsers for YAML and other markup languages if I remember right.

**Comment by GiovanniSalmeri** [comment]
Just another idea, **if** you are talking about a small amount of data (some numbers, for example), and **if** these data are not your bank credentials or the secret codes for beginning a nuclear war: you could simply transmit them with a GET request, for example `https://mysite.org/receivingpage/firstnumber:12345/secondnumber:6789/firstcode:aBcD/secondcode:eFgHi`

In page `receivingpage` you put a shortcut for example `[receiver]`: now in an extension `receiver.php` you can access your data with the Yellow API, i.e. `$page->getRequest("firstnumber")`, `$page->getRequest("secondnumber")` etc., and then save them, display them etc.

**Comment by PetersOtto** [comment]
Thank you both!
That's enough ideas! Sometimes I just think too complicated. FTP/SFTP did not occur to me, although I work with it almost every day. &#128533;
I will try both ideas. 

@GiovanniSalmeri My bank credentials may see everyone, there is nothing to get and my nuclear missiles I have sold years ago. &#128516;

**Comment by markseuffert** [comment, marked as answer]
Here are two alternatives for sending data to a website:

* Send a `POST` request via HTTP/HTTPS. You can send arbitrary data to a web server. This can be done in any programming language and at the command line. There are three [HTTP request methods](https://www.w3schools.com/tags/ref_httpmethods.asp) you should know: GET, POST and HEAD. The first one is for displaying pages and retrieving data from a web server. The second is for contact forms, uploading files and sending data to a web server. The third is for obtaining meta data and testing links. Datenstrom Yellow supports these HTTP request methods. Have a look at `getRequest()`, `isRequest()` in extensions and the API for developers. 

* Use the `rsync` command on Mac or Linux. You can sync files between a local computer and a remote web server. As soon as files are available in the file system, your website can show them. Maybe you can export your sensor data as image files, then you can show them right away. Syncing files is super easy if you already have SSH configured with key-based authentication. Here's an example command to sync the local folder `/shared/sensordata/` with a remote folder: 

  `rsync -avze ssh /shared/sensordata/ webserver.com:/var/www/website/media/images/sensordata/`

  It's such a useful tool. Have a look at this [tutorial video for rsync](https://www.youtube.com/watch?v=qE77MbDnljA).

Hope this helps and happy coding.

**Comment by PetersOtto** [comment]
Hello Mark

Thank you very much for the information! Because the Raspberry Pi Pico W is an 8 Euro microcontroller and not a real computer, the possibilities are limited. The `FTP/SFTP` solution should work, there is a library for it. But I like the `POST/GET` solution because I don't need passwords. I think the `rsync `solution will be difficult because the Pico has no operating system (only micropython).

But I now have enough ideas to continue. Thanks again. &#128522;

**Comment by annaesvensson** [comment]
Tip: You can let us know when your question has been answered, click on `Mark as answer`.
